src/bt.rs

```diff
@@ -0,0 +1,435 @@
+//! - Serializable: tree can be coverted to and from bytes.
+//!
+//! ## Creation
+//! Use [`BinaryTree::default`] to create a binary tree, note that there must be
+//! one value at the root node.
+//!
+//! ## Insertion
+//! Given a value and a binary path, [`BinaryTree::insert`] stores the value
+//! at the end of the path.
+//!
+//! ## Query
+//! Given a binary path, [`BinaryTree::get`] returns a reference to the value
+//! stored in the node located at the end of the path. The empty path returns
+//! the value at the root node. No value is returned if the end of the path
+//! is unreachable.
+//!
+//! ## Serialization
+//! A tree can be serializad to and from bytes using [`BinaryTree::encode`]
+//! and [`BinaryTree::decode_with_param`] functions, respectively. One-byte
+//! markers are used allowing to store sparse trees with a lower overhead.
+//!
+//! ## Example
+//! This binary tree can be created with the following code:
+//!
+//! ```txt
+//!               ()=@
+//!          /              \
+//!       (0)=a           (1)=b
+//!     /      \         /      \
+//! (00)=c  (01)=d   (10)=e  (11)=f
+//! ```
+//!
+//! use prio::bt::BinaryTree;
+//! use bitvec::{bits,prelude::Lsb0};
+//! let mut tree = BinaryTree::default();
+//! tree.insert(bits!(), '@');
+//! tree.insert(bits!(0), 'a');
+//! tree.insert(bits!(1), 'b');
+//! tree.insert(bits!(0, 0), 'c');
+//! tree.insert(bits!(0, 1), 'd');
+//! tree.insert(bits!(1, 0), 'e');
+//! tree.insert(bits!(1, 1), 'f');
+
+use core::fmt::Debug;
+use std::io::Cursor;
+
+use bitvec::slice::BitSlice;
+
+use crate::codec::{CodecError, Decode, Encode, ParameterizedDecode};
+
+/// Errors triggered by binary tree operations.
+#[derive(Debug, thiserror::Error)]
+#[non_exhaustive]
+pub enum BinaryTreeError<V> {
+    /// Error when inserting in a node with existing value.
+    #[error("node already contains a value")]
+    InsertNonEmptyNode(V),
+    /// Error when an operation cannot reach the node at the end of a path.
+    #[error("unreachable node in the tree")]
+    UnreachableNode(V),
+}
+
+/// Used to indicate a traversal path on the binary tree.
+pub type Path = BitSlice;
+
+type SubTree<V> = Option<Box<Node<V>>>;
+
+/// Represents a node of a binary tree.
+pub struct Node<V> {
+    value: V,
+    left: SubTree<V>,
+    right: SubTree<V>,
+}
+
+impl<V> Node<V> {
+    fn new(value: V) -> Self {
+        Self {
+            value,
+            left: None,
+            right: None,
+        }
+    }
+
+    /// Inserts the value at the end of the path.
+    ///
+    /// # Returns
+    /// - `Ok(())`, the node was inserted at the end of the path.
+    /// - `Err(InsertNonEmptyNode(value))`, the node already contains a value.
+    /// - `Err(UnreachablePath(value))`, the end of the path is unreachable.
+    ///
+    /// In the error cases, no insertion occurs and the value is returned to its owner.
+    pub fn insert(&mut self, path: &Path, value: V) -> Result<(), BinaryTreeError<V>> {
```

Comment:
I think this could use some more explanation, particularly that `Node::insert()` expects only part of a node's index, and particularly just enough bits to navigate from `self` to another node. (as opposed to `BinaryTree::insert()`, which takes the entire node index, since it is starting from the root) So, this would be a suffix of a prefix of a Mastic input ultimately.

Reply:
updated, hope is clearer now.